File: android-stery-print-bridge/app/src/main/java/com/stery/bluetoothprintbridge/ReceiptPayload.kt
package com.stery.bluetoothprintbridge

import org.json.JSONArray
import org.json.JSONObject

data class ReceiptItem(
    val name: String,
    val qty: Double,
    val unitPrice: Double,
    val lineTotal: Double,
)

data class ReceiptPayload(
    val storeName: String,
    val branch: String? = null,
    val receiptNo: String? = null,
    val cashier: String? = null,
    val customer: String? = null,
    val date: String? = null,
    val items: List<ReceiptItem> = emptyList(),
    val subtotal: Double? = null,
    val discount: Double? = null,
    val tax: Double? = null,
    val total: Double? = null,
    val paid: Double? = null,
    val change: Double? = null,
    val paymentMethod: String? = null,
    val footer: String? = null,
    val printLogo: Boolean = false,
    val duplicateLabel: String? = null,
    val note: String? = null,
) {
    companion object {
        fun fromJson(json: JSONObject): ReceiptPayload {
            val itemsArray = json.optJSONArray("items") ?: JSONArray()
            val items = buildList {
                for (i in 0 until itemsArray.length()) {
                    val item = itemsArray.optJSONObject(i) ?: continue
                    add(
                        ReceiptItem(
                            name = item.optString("name", "Item"),
                            qty = item.optDouble("qty", 1.0),
                            unitPrice = item.optDouble("unitPrice", 0.0),
                            lineTotal = item.optDouble("lineTotal", item.optDouble("total", 0.0)),
                        )
                    )
                }
            }

            return ReceiptPayload(
                storeName = json.optString("storeName", "STERY POS"),
                branch = json.optString("branch").takeIf { it.isNotBlank() },
                receiptNo = json.optString("receiptNo").takeIf { it.isNotBlank() },
                cashier = json.optString("cashier").takeIf { it.isNotBlank() },
                customer = json.optString("customer").takeIf { it.isNotBlank() },
                date = json.optString("date").takeIf { it.isNotBlank() },
                items = items,
                subtotal = json.optDoubleOrNull("subtotal"),
                discount = json.optDoubleOrNull("discount"),
                tax = json.optDoubleOrNull("tax"),
                total = json.optDoubleOrNull("total"),
                paid = json.optDoubleOrNull("paid"),
                change = json.optDoubleOrNull("change"),
                paymentMethod = json.optString("paymentMethod").takeIf { it.isNotBlank() },
                footer = json.optString("footer").takeIf { it.isNotBlank() },
                printLogo = json.optBoolean("printLogo", false),
                duplicateLabel = json.optString("duplicateLabel").takeIf { it.isNotBlank() },
                note = json.optString("note").takeIf { it.isNotBlank() },
            )
        }
    }
}

private fun JSONObject.optDoubleOrNull(name: String): Double? {
    return if (has(name) && !isNull(name)) optDouble(name, Double.NaN).takeIf { !it.isNaN() } else null
}

File: print-bridge/receiptPayload.js
function sanitizeText(value) {
  return String(value ?? "")
    .replace(/[\u0000-\u0008\u000B\u000C\u000E-\u001F\u007F]/g, "")
    .replace(/[\u2018\u2019\u201A\u201B]/g, "'")
    .replace(/[\u201C\u201D\u201E\u201F]/g, '"')
    .replace(/[\u2013\u2014]/g, "-")
    .replace(/\u00A0/g, " ")
    .replace(/\r\n/g, "\n")
    .replace(/\r/g, "\n")
    .trim();
}

function toNumber(value, fallback = 0) {
  if (value === null || value === undefined || value === "") return fallback;
  const cleaned = String(value).replace(/,/g, "").replace(/[^0-9.-]/g, "");
  const parsed = Number.parseFloat(cleaned);
  return Number.isFinite(parsed) ? parsed : fallback;
}

function normalizeReceiptItem(input) {
  const source = input && typeof input === "object" ? input : {};
  return {
    name: sanitizeText(source.name || "Item"),
    qty: toNumber(source.qty, 1),
    unitPrice: toNumber(source.unitPrice, 0),
    lineTotal: toNumber(source.lineTotal ?? source.total, 0),
  };
}

function normalizeReceiptPayload(input) {
  const source = input && typeof input === "object" ? input : {};
  const items = Array.isArray(source.items) ? source.items.map(normalizeReceiptItem) : [];

  return {
    storeName: sanitizeText(source.storeName || "STERY WHOLESALERS"),
    branch: sanitizeText(source.branch || ""),
    receiptNo: sanitizeText(source.receiptNo || ""),
    cashier: sanitizeText(source.cashier || ""),
    customer: sanitizeText(source.customer || ""),
    date: sanitizeText(source.date || ""),
    items,
    subtotal: source.subtotal === null || source.subtotal === undefined || source.subtotal === "" ? null : toNumber(source.subtotal, 0),
    discount: source.discount === null || source.discount === undefined || source.discount === "" ? null : toNumber(source.discount, 0),
    tax: source.tax === null || source.tax === undefined || source.tax === "" ? null : toNumber(source.tax, 0),
    total: source.total === null || source.total === undefined || source.total === "" ? null : toNumber(source.total, 0),
    paid: source.paid === null || source.paid === undefined || source.paid === "" ? null : toNumber(source.paid, 0),
    change: source.change === null || source.change === undefined || source.change === "" ? null : toNumber(source.change, 0),
    paymentMethod: sanitizeText(source.paymentMethod || ""),
    footer: sanitizeText(source.footer || "Thank you for shopping with us"),
    printLogo: Boolean(source.printLogo),
    duplicateLabel: sanitizeText(source.duplicateLabel || ""),
    note: sanitizeText(source.note || ""),
  };
}

module.exports = {
  normalizeReceiptPayload,
  sanitizeText,
  toNumber,
};
